pilot/pkg/xds/sds.go

```diff
@@ -134,39 +136,61 @@ func (s *SecretGen) Generate(proxy *model.Proxy, push *model.PushContext, w *mod
 			continue
 		}
 		regenerated++
-		// Fetch the appropriate cluster's secret, based on the credential type
-		var secretController credscontroller.Controller
-		switch sr.Type {
-		case credentials.KubernetesGatewaySecretType:
-			secretController = configClusterSecrets
-		default:
-			secretController = proxyClusterSecrets
+		res := s.generate(sr, configClusterSecrets, proxyClusterSecrets)
+		if res != nil {
+			s.cache.Add(sr, req, res)
+			results = append(results, res)
 		}
+	}
+	return results, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cached, cached+regenerated)}, nil
+}
 
-		isCAOnlySecret := strings.HasSuffix(sr.Name, securitymodel.SdsCaSuffix)
-		if isCAOnlySecret {
-			caCert, err := secretController.GetCaCert(sr.Name, sr.Namespace)
-			if err != nil {
-				pilotSDSCertificateErrors.Increment()
-				log.Warnf("failed to fetch ca certificate for %s: %v", sr.ResourceName, err)
-			} else {
-				res := toEnvoyCaSecret(sr.ResourceName, caCert)
-				results = append(results, res)
-				s.cache.Add(sr, req, res)
-			}
-		} else {
-			key, cert, err := secretController.GetKeyAndCert(sr.Name, sr.Namespace)
-			if err != nil {
+func (s *SecretGen) generate(sr SecretResource, configClusterSecrets, proxyClusterSecrets credscontroller.Controller) *discovery.Resource {
+	// Fetch the appropriate cluster's secret, based on the credential type
+	var secretController credscontroller.Controller
+	switch sr.Type {
+	case credentials.KubernetesGatewaySecretType:
+		secretController = configClusterSecrets
+	default:
+		secretController = proxyClusterSecrets
+	}
+
+	isCAOnlySecret := strings.HasSuffix(sr.Name, securitymodel.SdsCaSuffix)
+	if isCAOnlySecret {
+		caCert, err := secretController.GetCaCert(sr.Name, sr.Namespace)
+		if err != nil {
+			pilotSDSCertificateErrors.Increment()
+			log.Warnf("failed to fetch ca certificate for %s: %v", sr.ResourceName, err)
+			return nil
+		}
+		if features.VerifySDSCertificate {
+			roots := x509.NewCertPool()
+			ok := roots.AppendCertsFromPEM(caCert)
+			if !ok {
 				pilotSDSCertificateErrors.Increment()
-				log.Warnf("failed to fetch key and certificate for %s: %v", sr.ResourceName, err)
-			} else {
-				res := toEnvoyKeyCertSecret(sr.ResourceName, key, cert)
-				results = append(results, res)
-				s.cache.Add(sr, req, res)
+				log.Warnf("failed to parse root certificate %s", sr.ResourceName)
+				return nil
 			}
 		}
+		res := toEnvoyCaSecret(sr.ResourceName, caCert)
+		return res
 	}
-	return results, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cached, cached+regenerated)}, nil
+
+	key, cert, err := secretController.GetKeyAndCert(sr.Name, sr.Namespace)
+	if err != nil {
+		pilotSDSCertificateErrors.Increment()
+		log.Warnf("failed to fetch key and certificate for %s: %v", sr.ResourceName, err)
+		return nil
+	}
+	if features.VerifySDSCertificate {
+		if _, err := tls.X509KeyPair(cert, key); err != nil {
```

Comment:
same `ParseCertificate` comment?

Reply:
Well might need a little more I guess, there is a lot of validation in `X509KeyPair` and not a lot of waste. But I think CertPool is not very cheap

Reply:
Will have a try and do a comparison